src/main/java/net/javadiscord/javabot/listener/HugListener.java

```diff
@@ -55,40 +59,49 @@ public void onMessageReceived(@Nonnull MessageReceivedEvent event) {
 		if (tc == null) {
 			return;
 		}
-		final TextChannel textChannel = tc;
 		String content = event.getMessage().getContentRaw();
-		String lowerCaseContent = content.toLowerCase();
-		if (lowerCaseContent.contains("fuck")) {
+		if (FUCKER.matcher(content).find()) {
 			long threadId = event.isFromThread() ? event.getChannel().getIdLong() : 0;
-			StringBuilder sb = new StringBuilder(content.length());
-			int index = 0;
-			int indexBkp = index;
-			while ((index = lowerCaseContent.indexOf("fuck", index)) != -1) {
-				sb.append(content.substring(indexBkp, index));
-				sb.append(loadHug(content, index));
-				indexBkp = index++ + 4;
-				if (content.length() >= indexBkp + 3 && "ing".equals(lowerCaseContent.substring(indexBkp, indexBkp + 3))) {
-					sb.append(copyCase(content, indexBkp-1, 'g'));
-					sb.append(content.substring(indexBkp, indexBkp + 3));
-					index+=3;
-					indexBkp+=3;
-				}
-			}
-
-			sb.append(content.substring(indexBkp));
-			WebhookUtil.ensureWebhookExists(textChannel,
-					wh -> sendWebhookMessage(wh, event.getMessage(), sb.toString(), threadId),
+			WebhookUtil.ensureWebhookExists(tc,
+					wh -> sendWebhookMessage(wh, event.getMessage(), replaceFucks(content), threadId),
 					e -> ExceptionLogger.capture(e, getClass().getSimpleName()));
 		}
 	}
 
-	private String loadHug(String originalText, int startIndex) {
-		return copyCase(originalText, startIndex, 'h') + ""
-				+ copyCase(originalText, startIndex + 1, 'u') + ""
-				+ copyCase(originalText, startIndex + 3, 'g');
+	private static String processHug(String originalText) {
+		// FucK -> HuG, FuCk -> Hug
+		return String.valueOf(copyCase(originalText, 0, 'h'))
+			+ copyCase(originalText, 1, 'u')
+			+ copyCase(originalText, 3, 'g');
+	}
+
+	private static String replaceFucks(String str) {
+		return FUCKER.matcher(str).replaceAll(matchResult -> {
+			String theFuck = matchResult.group(1);
+			String suffix = Objects.requireNonNullElse(matchResult.group(2), "");
+			String processedSuffix = switch(suffix.toLowerCase()) {
+				case "er", "ing" -> copyCase(suffix, 0, 'g') + suffix; // fucking, fucker
+				case "" -> ""; // just fuck
+				default -> copyCase(suffix, "g".repeat(suffix.length())); // fuckkkkk...
+			};
+			return processHug(theFuck) + processedSuffix;
+		});
+	}
+
+	private static String copyCase(String source, String toChange) {
+		if (source.length() != toChange.length()) throw new IllegalArgumentException("lengths differ");
+		StringBuilder sb = new StringBuilder();
+		for (int i = 0; i < source.length(); i++) {
```

Comment:
I think it would make more sense to just replace the loop body with `sb.append(copyCase(source, i, toChange.charAt(i)));`